### PR TITLE
linux capabilities: normalization and auditbeat process support

### PR DESCRIPTION
**Pull request by haesbaert**
## Proposed commit message

Add process capabilities to auditbeat/system and normalize the existing capabilities uses in beats.

The main goal is to add `process.thread.capabilities.{permitted, effective}` only, but since there are a couple of other uses of capabilities in beats already, it makes sense to standarize them into a common way to generate the required ECS fields.

I noticed that on root processes, where all capabilities are set, we end up generating a ton of output, so I proposed to condense them into CAP_ALL [here](https://github.com/elastic/security-team/issues/4375#issuecomment-1846197475).
~~This change implements said CAP_ALL and also makes sure the other beats end up outputting CAP_ALL as well by using `libbeat/common/capabilities/capabilities_*`.~~ CAP_ALL removed from this PR, to be discussed in the future.

The original idea was to rely on go-sysinfo that parses `/proc`, and while that works fine, we need more interfaces: like being able to get capabilities from an uint64 that comes from a kprobe for example, so instead I'm using the already imported `cap` module that is shipped with libcap(3) that retrieves process capabilities via capget(2), it's also maintained by libcap people so hopefully we don't have to care about maintaining our own capability table in the future. Additionally it also handles old kernels that have a slightly different ABI.

~~`cap` does runtime discovery of the maximum capability set of the host (`cap.MaxBits()`) and that is the criteria to issuing a `CAP_ALL`, note that if the capability being converted contains more bits (say we're using an old `cap` module in a newer kernel), the capability will still be exported but as "CAP_42, 43...", and CAP_ALL will be suppressed, I believe this is way less confusing than CAP_ALL+CAP_42.~~ CAP_ALL removed from this PR, to be discussed in the future.

Beats seem to be tracking ECS 8.0.0-dev, these new fields appeared in 8.10 so I've added them to the auditbeat fields in order to pass the integration tests, otherwise it complains about undocumented fields. 

I've added capget(2) to all beats syscomp policy, this is because `cap.GetProc()` is [written with the assumption that you can never get EPERM for the current process](https://git.kernel.org/pub/scm/libs/libcap/libcap.git/tree/cap/cap.go#n359).

I've also raised the point that there is a canonical text representation for capabilities https://linux.die.net/man/3/cap_to_text
```
$ ./a.out "cap_chown=p cap_chown+e"
caps_to_text() returned "= cap_chown+ep"
```
And I wonder if it's too late and/or makes any sense to adopt something like that in ECS.

## Checklist

- [] ~~My code follows the style guidelines of this project~~
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

I've added tests and converted the journald ones to use the same backing, and verified it still works. I've also tested teh journald output on an elastic deployment to be sure.

There is an obvious TOCTOU in fetching the capabilities since when we attempt to retrieve them the process might be gone (ESRCH), in this case we set `process.Error` if not already set, I wonder if it makes sense to not error out in this case and just mark CAP_UNKNOWN or whatever, I'm not familiar enough with beats to judge this.

## How to test this PR locally

auditbeat configuration:
```
auditbeat.modules:

- module: system
  datasets:
    - process
#    - socket
  period: 1m # The frequency at which the datasets check for changes


  # How often datasets send state updates with the
  # current state of the system (e.g. all currently
  # running processes, all open sockets).
  state.period: 5m

output.console:
  pretty: true
```
```
$ sudo auditbeat -c /tmp/auditbeat.yml | grep -C100 CAP_
{
  "@timestamp": "2023-12-05T19:34:54.425Z",
  "@metadata": {
    "beat": "auditbeat",
    "type": "_doc",
    "version": "8.12.0"
  },
  "process": {
    "thread": {
      "capabilities": {
        "effective": [
          "CAP_DAC_READ_SEARCH",
          "CAP_SYS_RESOURCE"
        ],
        "permitted": [
          "CAP_DAC_READ_SEARCH",
          "CAP_SYS_RESOURCE"
        ]
      }
    },
    "entity_id": "DADEDQU03GoDNhc1",
    "pid": 2841325,
    "start": "2023-12-05T19:32:53.180Z",
    "args": [
      "systemd-userwork: waiting..."
    ],
....
```

## Related issues

 - Relates https://github.com/elastic/beats/issues/36404
 - Relates https://github.com/elastic/security-team/issues/4375

## Screenshots

CAP_ALL
some caps
no caps

### CODE REVIEW

File: libbeat/common/capabilities/capabilities_linux.go [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+	errInvalidCapability = errors.New("invalid capability")
+
+	// Capabilities are linux only and this is returned on other
+	// systems for all the public functions. There is a generic
+	// errors.ErrUnsupported present in golang 1.21, but we still
+	// support 1.20.
+	ErrUnsupported = errors.New("capabilities are only supported in linux")
+
+	// The mask when all known capabilities are set.
+	allMask = (uint64(1) << uint64(cap.MaxBits())) - 1
```
Comment (Tacklebox) [inline libbeat/common/capabilities/capabilities_linux.go:42]: Should this not be something like 

```go
allMask = (uint64(1) << uint64(len(ecsNames)) - 1
```
Otherwise if you are on a newer kernel with unnamed capabilities you will end up using a valid capability bit to indicate that something has every capability?

Comment (haesbaert) [reply libbeat/common/capabilities/capabilities_linux.go:42]: You're right that "CAP_ALL" is host dependent, and the allMask reflects what CAP_ALL is from the host perspective.
I'm not sure I follow you, "if you are on a *newer* kernel with unnamed capabilities", did you mean an *older* kernel?
What will happen now is "if this is not my idea of CAP_ALL, make list with each individual capability, and if I don't know the name of some, call them CAP_NUMBER". Does that make sense?

File: libbeat/common/capabilities/capabilities_linux.go [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+		return []string{"CAP_ALL"}, nil
+	}
+
+	sl := make([]string, 0, bits.OnesCount64(w))
+	for i := 0; w != 0; i++ {
+		if w&1 != 0 {
+			s, err := toECS(i)
+			// impossible since MaxBits <= 64
+			if err != nil {
+				return []string{}, err
```
[the remark this answers is not in the document]
Comment (haesbaert) [reply libbeat/common/capabilities/capabilities_linux.go:188]: Aha after some googling I see returning `nil` there is more idiomatic, should I change?

File: x-pack/auditbeat/module/system/process/process.go [hunk 3]
```diff
@@ -488,6 +499,14 @@ func (ms *MetricSet) getProcesses() ([]*Process, error) {
 			process.UserInfo = &userInfo
 		}
 
+		process.CapEffective, err = capabilities.FromPid(capabilities.Effective, pInfo.PID)
+		if err != nil && !errors.Is(err, capabilities.ErrUnsupported) && process.Error == nil {
+			process.Error = err
+		}
+		process.CapPermitted, err = capabilities.FromPid(capabilities.Permitted, pInfo.PID)
+		if err != nil && !errors.Is(err, capabilities.ErrUnsupported) && process.Error == nil {
+			process.Error = err
```
[the remark this answers is not in the document]
Comment (haesbaert) [reply x-pack/auditbeat/module/system/process/process.go:508]: True, but two things: code moves around (people copy and paste stuff); and symmetry is nice :). So I'd vote for leaving it.

File: libbeat/common/capabilities/capabilities_linux_test.go [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+	assertHasCap(t, sl, "CAP_DAC_OVERRIDE")
+	assertHasCap(t, sl, "CAP_DAC_READ_SEARCH")
+	assertHasCap(t, sl, "CAP_FOWNER")
+	assertHasCap(t, sl, "CAP_FSETID")
+	assertHasCap(t, sl, "CAP_KILL")
+	assertHasCap(t, sl, "CAP_SETGID")
+	assertHasCap(t, sl, "CAP_SYS_MODULE")
+	assertHasCap(t, sl, "CAP_SYS_RAWIO")
+	assertHasCap(t, sl, "CAP_IPC_LOCK")
+	assertHasCap(t, sl, "CAP_MAC_OVERRIDE")
```
Comment (efd6) [inline libbeat/common/capabilities/capabilities_linux_test.go:76]: Tables make this clearer.
```suggestion
	for _, cap := range []string{
		"CAP_CHOWN",
		"CAP_DAC_OVERRIDE",
		"CAP_DAC_READ_SEARCH",
		"CAP_FOWNER",
		"CAP_FSETID",
		"CAP_KILL",
		"CAP_SETGID",
		"CAP_SYS_MODULE",
		"CAP_SYS_RAWIO",
		"CAP_IPC_LOCK",
		"CAP_MAC_OVERRIDE",
	} {
		assertHasCap(t, sl, cap)
	}
```

File: libbeat/common/capabilities/capabilities_linux_test.go [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+func assertHasCap(t *testing.T, sl []string, s string) {
+	var found int
+
+	for _, s2 := range sl {
+		if s2 == s {
+			found++
+		}
+	}
+
+	assert.Equal(t, found, 1)
```
Comment (efd6) [inline libbeat/common/capabilities/capabilities_linux_test.go:94]:
```suggestion
	assert.Equal(t, found, 1, s)
```

File: libbeat/common/capabilities/capabilities_other.go [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+const (
+	// Meaningless on "not linux".
+	Effective = Flag(0)
+	// Meaningless on "not linux".
+	Permitted = Flag(1)
+)
+
+// ErrUnsupported on "not linux".
+func FromPid(flag Flag, pid int) ([]string, error) {
+	return []string{}, ErrUnsupported
```
Comment (efd6) [inline libbeat/common/capabilities/capabilities_other.go:39]:
```suggestion
	return nil, ErrUnsupported
```
and below

File: x-pack/auditbeat/module/system/process/process.go [hunk 7]
```diff
@@ -376,6 +380,13 @@ func (ms *MetricSet) processEvent(process *Process, eventType string, action eve
 		event.RootFields.Put("user.group.name", process.Group.Name)
 	}
 
+	if len(process.CapEffective) > 0 {
```
Comment (efd6) [inline x-pack/auditbeat/module/system/process/process.go:383]:
```suggestion
	if len(process.CapEffective) != 0 {
```
(`cap` and `len` are spec'd to return positive `int` values)
also below

File: libbeat/common/capabilities/capabilities_linux.go [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+		all.SetFlag(cap.Inheritable, true, cap.Value(i))
+	}
+
+	return func(flag Flag, set *cap.Set) (bool, error) {
+		return isEqual(flag, set, all)
+	}
+}
+
+// Like isAll(), but for the empty set, here for symmetry.
+var isEmpty = makeIsEmpty()
```
[the remark this answers is not in the document]
Comment (haesbaert) [reply libbeat/common/capabilities/capabilities_linux.go:119]: I'd have to `s/emtpy/cap.NewSet()`, which allocates a set per call, I'd be fine with it but since there is already the same scheme with `isAll`, I'd leave it is.

File: filebeat/input/journald/pkg/journalfield/conv_expand_test.go [hunk 9]
```diff
@@ -121,47 +123,7 @@ var expandCapabilitiesTests = []struct {
...
-							"CAP_MAC_OVERRIDE",
-							"CAP_MAC_ADMIN",
-							"CAP_SYSLOG",
-							"CAP_WAKE_ALARM",
-							"CAP_BLOCK_SUSPEND",
-							"CAP_AUDIT_READ",
-							"CAP_PERFMON",
-							"CAP_BPF",
-							"CAP_CHECKPOINT_RESTORE",
+							"CAP_ALL",
```
[the remark this answers is not in the document]
Comment (haesbaert) [reply filebeat/input/journald/pkg/journalfield/conv_expand_test.go:126]: Indeed, perhaps we should convey this better to the people managing ECS. I think this would be fine since even though they would reflect different things, they would still "semantically" reflect the same as in this process could do all it wanted.

File: libbeat/common/capabilities/capabilities_linux.go [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+		if err != nil {
+			return []string{}, err
+		}
+		if !v {
+			continue
+		}
+		s, err := toECS(i)
+		// impossible since MaxBits <= 64
+		if err != nil {
+			return []string{}, err
```
[the remark this answers is not in the document]
Comment (haesbaert) [reply libbeat/common/capabilities/capabilities_linux.go]: I'm never completely confident of anything :). If this was new repository with code that doesn't have users I'd panic.

File: libbeat/common/capabilities/capabilities_linux.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (haesbaert) [reply libbeat/common/capabilities/capabilities_linux.go]: all changed

File: libbeat/common/capabilities/capabilities_linux.go [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+const (
+	// aka CapEff
+	Effective = cap.Effective
+	// aka CapPrm
+	Permitted = cap.Permitted
+)
+
+// True if sets are equal for the given flag/vector, errors out in
+// case any of the sets is malformed.
+func isEqual(flag Flag, a *cap.Set, b *cap.Set) (bool, error) {
```
[the remark this answers is not in the document]
Comment (haesbaert) [reply libbeat/common/capabilities/capabilities_linux.go]: Done, leme know if this isn't what you meant.

File: filebeat/input/journald/pkg/journalfield/conv_expand_test.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (efd6) [reply filebeat/input/journald/pkg/journalfield/conv_expand_test.go:126]: I'm not convince by that. If there were a flag that said ALL, then I would agree, but the ALLness is determined by comparison to the set of valid flags that were present at the time of the writing of the document. Essentially, the semantics are inferred contextually and then rendered concrete without reference to the context. I don't think this is safe. Note that ECS doesn't mention CAP_ALL explicitly, and certainly doesn't recommend that the set be collapsed to it in the case that all capabilities are present, so I don't see that we HAVE TO do this.

[the remark this answers is not in the document]
Comment (efd6) [reply filebeat/input/journald/pkg/journalfield/conv_expand_test.go:126]: To expand on this, we COULD do this if we also keep all the original flags as well. This enriches the information in the document by saying, "this is the set of capabilities that are present *and that was the complete set available at the time*".

[the remark this answers is not in the document]
Comment (haesbaert) [reply filebeat/input/journald/pkg/journalfield/conv_expand_test.go:126]: > I'm not convince by that. If there were a flag that said ALL, then I would agree, but the ALLness is determined by **comparison to the set of valid flags that were present at the time of the writing of the document.** Essentially, the semantics are inferred contextually and then rendered concrete without reference to the context. I don't think this is safe. Note that ECS doesn't mention CAP_ALL explicitly, and certainly doesn't recommend that the set be collapsed to it in the case that all capabilities are present, so I don't see that we HAVE TO do this.

I don't think that is quite the case, the comparison of the valid flags **are not the ones present at the time of writing the document**, they are what the kernel determines to be "all the capabilities it supports", cap.MaxBits() is discovered at [runtime](https://git.kernel.org/pub/scm/libs/libcap/libcap.git/tree/cap/cap.go#n276).
"MaxBits returns the number of kernel-named capabilities discovered at runtime in the current system."
So if the kernel adds another capability, even if we don't update the `cap` pkg, MaxBits will be bumped and CAP_ALL will reflect all the capabilities the new kernel supports.

If we don't update `cap`, the only thing we miss is the name of the new capability (which will be reflected as CAP_NUMBER in the case of not CAP_ALL).

Regarding it not being documented in ECS, this has just been suggested [here](https://github.com/elastic/security-team/issues/4375#issuecomment-1846197475), so I assume documentation will follow.

Regarding keeping both sets, the whole point is to shrink the amount of flags, since every root process will spill the 41 capabilities now.

File: libbeat/common/capabilities/capabilities_linux_test.go [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+	sl, err := FromString("0", 16)
+	assert.Nil(t, err)
+	assert.Equal(t, len(sl), 0)
+
+	sl, err = FromUint64(0)
+	assert.Nil(t, err)
+	assert.Equal(t, len(sl), 0)
+
+	// assumes non root has no capabilities
+	if os.Geteuid() != 0 {
```
[the remark this answers is not in the document]
Comment (haesbaert) [reply libbeat/common/capabilities/capabilities_linux_test.go:39]: Indeed, you can still have capabilities as a non-root user, I've added the comment to reflect that, I guess it's just rare enough that we could live with it.
I can't think of a good way to mock it, one way would be to drop all capabilities that are "Effective & Permitted", then test Effective for empty and restore the old set, but this is too impure and doesn't sound very appealing to me.
I'd either leave it and improve the comment or just remove it all together, let me know what you think.

File: libbeat/common/capabilities/capabilities_linux.go [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+		return []string{}, err
+	}
+	if all {
+		return []string{"CAP_ALL"}, nil
+	}
+
+	var sl []string
+	for i := 0; i < int(cap.MaxBits()); i++ {
+		c := cap.Value(i)
+		v, err := set.GetFlag(flag, c)
```
[the remark this answers is not in the document]
Comment (haesbaert) [reply libbeat/common/capabilities/capabilities_linux.go]: pushed the suggestion

File: filebeat/input/journald/pkg/journalfield/conv_expand_test.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (haesbaert) [reply filebeat/input/journald/pkg/journalfield/conv_expand_test.go:126]: Answering myself, the runtime discovery _does_ what we want, it looks for the first bit that returns EINVAL, not the first bit that is not set. Tested on docker with a smaller bounding set.
```
capget({version=_LINUX_CAPABILITY_VERSION_3, pid=0}, NULL) = 0
prctl(PR_CAPBSET_READ, CAP_MAC_OVERRIDE) = 0
prctl(PR_CAPBSET_READ, 0x30 /* CAP_??? */) = -1 EINVAL (Invalid argument)
prctl(PR_CAPBSET_READ, CAP_CHECKPOINT_RESTORE) = 0
prctl(PR_CAPBSET_READ, 0x2c /* CAP_??? */) = -1 EINVAL (Invalid argument)
prctl(PR_CAPBSET_READ, 0x2a /* CAP_??? */) = -1 EINVAL (Invalid argument)
prctl(PR_CAPBSET_READ, 0x29 /* CAP_??? */) = -1 EINVAL (Invalid argument)
write(1, "maxbits=41\n", 11maxbits=41
```
and `cat /proc/self/status` is:
```
CapInh: 0000000000000000
CapPrm: 00000000a80425fb
CapEff: 00000000a80425fb
CapBnd: 00000000a80425fb
CapAmb: 0000000000000000
```

In other words, MaxBits is correct regardless of the bounding set (which containers run with a limited one), meaning we can trust that "CAP_ALL is every capability the host can have".

**Comment by haesbaert** [comment]
> @haesbaert is it ready for another review?

Yes, nothing considerable changed, I've pinged @andrewkroh to get his input on the CAP_ALL semantics. I've also spoken to Dan about just removing CAP_ALL in order to make this PR go forward and then bring it in a subsequent PR.
I've also got the bits for the process processor ready once this is in.

File: CHANGELOG.next.asciidoc [hunk 14]
```diff
@@ -132,6 +132,9 @@ Setting environmental variable ELASTIC_NETINFO:false in Elastic Agent pod will d
 
 *Auditbeat*
 
+- Add `ignore_errors` option to audit module. {issue}15768[15768] {pull}36851[36851]
+- Fix copy arguments for strict aligned architectures. {pull}36976[36976]
```
Comment (mmat11) [inline CHANGELOG.next.asciidoc:136]:
```suggestion
```
I think the first two are not part of this PR, right?

**Comment by haesbaert** [comment]
I have removed the CAP_ALL optimization the interesting of expediting this PR, I think it makes more sense to get this in on an easy to review PR and then discuss CAP_ALL later, as I also have the process processor implementation to get in after this.
Also minor improvements:
 - beats is in Go 1.21 so we can use the generic errors.ErrNotSupported instead of rolling our own
 - stop trying to get capabilities for kernel process as we're just about to discard them from the set

File: CHANGELOG.next.asciidoc (hunk 14 above)
Comment (haesbaert) [reply CHANGELOG.next.asciidoc:136]: Good catch! That probably snuck in in one of my merges

**Review by Tacklebox — APPROVED** [review]
Looks 👍

**Review by efd6 — APPROVED** [review]

**Review by leehinman — APPROVED** [review]
